### PR TITLE
maintainers/scripts/haskell/hydra-report.hs: init

### PR DESCRIPTION
**Pull request by maralorn**
An example output of running `ping-maintainers` can be found in the top-level comment at #121627.

As often when writing a Haskell script (especially with JSON and printing) quickly the line number grows. I hope that’s okay.

I welcome any feedback!

### CODE REVIEW

**Comment by maralorn** [comment]
@sorki Sorry for using you as a guinea pig. But would you mind telling me if highlighting you in an edit to #121627 did trigger a github notification for that PR?

**Comment by sorki** [comment]
> @sorki Sorry for using you as a guinea pig. But would you mind telling me if highlighting you in an edit to #121627 did trigger a github notification for that PR?

No problem! Yes it did, thanks for that. Now I wonder if we can list multiple maintainers for one package.

**Comment by maralorn** [comment]
> Now I wonder if we can list multiple maintainers for one package.

If I am not terribly mistaken we totally can and the script does. It’s just that at the moment most packages only have one maintainer.

File: maintainers/scripts/haskell/hydra-report.hs
Comment (sternenseemann) [inline maintainers/scripts/haskell/hydra-report.hs:127]: I think this should retry on failure a few times. Hydra has the habit of randomly 500ing on you.

File: maintainers/scripts/haskell/hydra-report.hs
Comment (sternenseemann) [inline maintainers/scripts/haskell/hydra-report.hs:156]: Note that I'm still planning to enable `aarch64` and `x86_64-darwin`.

File: maintainers/scripts/haskell/hydra-report.hs
Comment (sternenseemann) [inline maintainers/scripts/haskell/hydra-report.hs:142]: This means we can't ping maintainers of top-level packages who have a maintainer? Some packages use the `haskellPackages` infrastructure, but have a dedicated nix expression outside of `haskellPackages`.

File: maintainers/scripts/haskell/hydra-report.hs
Comment (sternenseemann) [inline maintainers/scripts/haskell/hydra-report.hs:210]:
```suggestion
            <> ["*Report generated with [maintainers/scripts/haskell/hydra-report.hs](https://github.com/NixOS/nixpkgs/blob/haskell-updates/maintainers/scripts/haskell/hydra-report.hs)*"]
```

File: maintainers/scripts/haskell/hydra-report.hs
Comment (sternenseemann) [inline maintainers/scripts/haskell/hydra-report.hs:114]: Maybe we should save this in `XDG_CACHE_HOME` (fallback to `$HOME/.cache`).

File: maintainers/scripts/haskell/hydra-report.hs
Comment (maralorn) [reply maintainers/scripts/haskell/hydra-report.hs:127]: Interestingly this has never failed me until now.

Actually I realize now, what we want is to keep one connection open and query via that one. I have no clue if req does that. Need to look into that.

But retrying is certainly a good idea.

File: maintainers/scripts/haskell/hydra-report.hs
Comment (maralorn) [reply maintainers/scripts/haskell/hydra-report.hs:156]: Yeah, I know. This script will at the moment just throw all the jobs from other platforms into "other jobs". Which is okay for now, maybe? Anyways I am not sure what the best desired behavior in the presence of multiple platforms is. Probably we want to somehow group builds for one package together? But the usability will depend a lot on how many failures we will experience.

File: maintainers/scripts/haskell/hydra-report.hs
Comment (maralorn) [reply maintainers/scripts/haskell/hydra-report.hs:142]: Yeas, right now it’s that way. I still think this is a huge win. I briefly tried to get the expression to include maintainers for top-level packages as well, but I met problems there. (Also it would probably make the script still slower. But I think it would be still okay.)

File: maintainers/scripts/haskell/hydra-report.hs
Comment (maralorn) [reply maintainers/scripts/haskell/hydra-report.hs:210]: oops

File: maintainers/scripts/haskell/hydra-report.hs
Comment (maralorn) [reply maintainers/scripts/haskell/hydra-report.hs:114]: Good idea. Allthough that increases the dependency footprint of the script, I will do it.

File: maintainers/scripts/haskell/hydra-report.hs
Comment (maralorn) [reply maintainers/scripts/haskell/hydra-report.hs:142]: Oh, btw. This goes actually back to the question I asked you in discord. The most principled way to do this would be to write a nix-expression that takes the job list from our test job iterates over that and queries the maintainers for that. Than we can be sure that we know the maintainers for every job.

Comment (sternenseemann) [reply maintainers/scripts/haskell/hydra-report.hs:142]: You can easily achieve that by running `hydra-eval-jobs` and using its json output. It only returns emails for the maintainers field however, but this should be enough to find the github name in most cases.

File: maintainers/scripts/haskell/hydra-report.hs
Comment (sternenseemann) [reply maintainers/scripts/haskell/hydra-report.hs:114]: I don't think pulling in a xdg library is necessary really, you can just have two calls to `getEnv`.

File: maintainers/scripts/haskell/hydra-report.hs
Comment (maralorn) [reply maintainers/scripts/haskell/hydra-report.hs:142]: Aaah, great!

File: maintainers/scripts/haskell/hydra-report.hs
Comment (maralorn) [reply maintainers/scripts/haskell/hydra-report.hs:126]: Okay. I have read the documentation of `req` and apparently it just has really good defaults. It already does connection sharing and retrying under the hood.

File: maintainers/scripts/haskell/hydra-report.hs
Comment (maralorn) [reply maintainers/scripts/haskell/hydra-report.hs]: The updated version of the script takes this into account.

File: maintainers/scripts/haskell/hydra-report.hs
Comment (maralorn) [reply maintainers/scripts/haskell/hydra-report.hs]: Done, we now find maintainers for all jobs.

File: maintainers/scripts/haskell/hydra-report.hs
Comment (maralorn) [reply maintainers/scripts/haskell/hydra-report.hs]: directory exports xdgDirectory and is bundled with ghc. So everything great here!

**Comment by maralorn** [comment]
I am very uncertain about the actual printed output but I think the current state is at least not worse than not having a report. Still very open to improvement suggestions.

**Comment by maralorn** [comment]
I think I have addressed all concerns.

**Comment by cdepillabout** [comment]
I tried running this locally and I ended up with an error about JSON from Hydra not being able to be decoded:

```console
$ ./maintainers/scripts/haskell/hydra-report.hs get-report
hydra-report.hs: JsonHttpException "Error in $.buildstatus: parsing Int failed, expected Number, but encountered Null"
```

I'm currently running this script against a Hydra that hasn't completed all the builds for the evaluation of the jobset, so it makes sense that `Build.buildstatus` would be `null`:

image

I've fixed this locally.  I'm currently running `hydra-report.hs get-report`.  When it successfully finishes, I'll push my fix for this.

**Comment by cdepillabout** [comment]
I tried running the script again but it failed with a timeout again, similar to above.  It failed around the 12,100th request or so.  It looks like some sort of retry logic might be necessary for me to be able to have the script finish running.

I timed it this time, and 12100 requests took around 1.5 hours.

**Comment by maralorn** [comment]
First thanks for the effort! Both changes seem very sensible.

Oh, man. Why 1.5h? For me a successful run of the script takes 33 minutes, what already seems much to high.
So the issue here is this. `req` already has a retry logic. According to [the documentation](https://hackage.haskell.org/package/retry-0.8.1.2/docs/Control-Retry.html#t:RetryPolicyM)
it retries 5 times with 50ms gaps.
What retry policy do you recommend?

I have the feeling a good additional idea would be to cache builds locally on disk so that you can a) resume fetching later without loosing progress and b) don‘t need to refetch all builds when fetching a new evaluation.

**Comment by maralorn** [comment]
Oh, and I think aborting on unfinished jobs is not the best way to go because those could be e.g. all aarch64 while x86 could already be actionable.

**Comment by cdepillabout** [comment]
@maralorn Thanks for responding.

> My best explanation for your much increased request time is network delay (I guess hydra is in Europe?)

You know, this is a good explanation.  This is probably the reason it takes so much longer for me, as well as the reason that I'm seeing timeouts when you guys are not.

Let me try to locally bump the timeout from 30s to 1min and see if I still get these timeouts.

If this works, then hopefully we won't have to change the retry policy.

> I have the feeling a good additional idea would be to cache builds locally on disk so that you can a) resume fetching later without loosing progress and b) don‘t need to refetch all builds when fetching a new evaluation.
>
> While I dread the feature creep and additional logic going into this.

I completely agree with both of these points.

> But if this is really so much network bound we should consider running a few requests concurrently.

Yeah, this might be a easy way to speed up the script.  Although like you said above, it is hard to know how much we are able to hammer Hydra until we start causing problems for other people.

> I think aborting on unfinished jobs is not the best way to go because those could be e.g. all aarch64 while x86 could already be actionable.

Ah, that is a good point.

I wasn't actually following the decision very closely about enabling the darwin and aarch64 builds, so I don't really understand what the effects would be of running this script in a state where Hydra has finished all the x86 builds, but has not yet finished all the aarch64 builds.

I meant to ask about this at some point, but as a maintainer, I'm not actually sure what my responsibility would be for things that fail to build on aarch64, but succeed on x86.  Do we have some script that automatically marks `unsupported-platforms` for things that build on x86, but fail to build on aarch64?

**Comment by maralorn** [comment]
> I wasn't actually following the decision very closely about enabling the darwin and aarch64 builds, so I don't really understand what the effects would be of running this script in a state where Hydra has finished all the x86 builds, but has not yet finished all the aarch64 builds. Would the `print-maintainers` subcommand from `hydra-report.hs` still make sense? I guess the `mark-broken-list` would still make sense though, right (since it is only currently looking at x86)?
> 
> I meant to ask about this at some point, but as a maintainer, I'm not actually sure what my responsibility would be for things that fail to build on aarch64, but succeed on x86. Do we have some script that automatically marks `unsupported-platforms` for things that build on x86, but fail to build on aarch64?

Nope. Nothing in that department has really been decided. My default until now is: Let’s proceed with the same policies as before but with more information. (i.e. we are welcome to fix darwin and arm, but allowed to ignore it.) `mark-broken-list` focuses on x86 right now so it’s not affected, also running `print-maintainers` with unfinished darwin jobs is not a big issue. There will just be :hourglass_flowing_sand: at the point indicating that job.

Of course we could consider doing this differently. We could e.g. only mark a build broken if it is broken on all platforms and remove supported platforms if it is only broken on some. Then we could proceed and propagate the dont-distribute state based on the platform. A lot of that is probably automatable but I am hesitant to put it on my todo-list.

**Comment by maralorn** [comment]
@cdepillabout I have an engineering question about the caching. Maybe you know a best practice.

The worst solution I can think of is having "Map BuildID Build" and on every new downloaded Build I read it from disk via "FromJson" insert the build with "ToJson" that way I write the file 18000 times.

Do I want an sqlite db for this?

The best solution I can think of would be this: Fetch buildids from hydra and read in builds from disk, where they live as on object per line.
Now open the file to write and immediately write back all builds that we had already downloaded and need in this new evaluation. After that start the downloads and append every build as soon as we have it. Since we do this in a withFile context any timeout or exception would probably not lead to loosing builds.

**Comment by sternenseemann** [comment]
I'm getting the impression that we maybe should add a means to query evry build status for a jobset at once instead of looking for some gains to reduce the run time of this script :| I could do it in theory as I have some experience with Perl, but not sure when I'll get to it.

> I meant to ask about this at some point, but as a maintainer, I'm not actually sure what my responsibility would be for things that fail to build on aarch64, but succeed on x86. Do we have some script that automatically marks unsupported-platforms for things that build on x86, but fail to build on aarch64?

I would hold up on this for a few weeks at the very least. I am currently actively working on improving the situation for those platforms and it seems to me that the effort of checking the build failures manually and then either fixing them or marking the package as unsupported is reasonable, but will lead to a much higher quality outcome than just assuming that packages don't work if they don't build out of the box.

Especially darwin needs a bit of manual intervention for packages that should build fine.

Adding darwin and aarch64 only added little under 400 new failures which is something I'm confident can be addressed manually.

File: maintainers/scripts/haskell/hydra-report.hs
Comment (sternenseemann) [inline maintainers/scripts/haskell/hydra-report.hs:119]: You want to do this differently:

* First you want to request `https://hydra.nixos.org/jobset/nixpkgs/haskell-updates/evals` to get the most recent evaluation
* Then `https://hydra.nixos.org/eval/<id>/builds` to get all builds and their status.

The second step is very slow, but much faster than what you have now and also probably better on hydra resources

**Review by sternenseemann — CHANGES_REQUESTED** [review]

**Comment by sternenseemann** [comment]
Additionally, I wonder since our main objective is getting the status of maintained packages, we should probably just check on the state of the `maintained` aggregate job instead. That would also save us from evaluating stuff locally.

File: maintainers/scripts/haskell/hydra-report.hs
Comment (sternenseemann) [reply maintainers/scripts/haskell/hydra-report.hs:119]: Still this is very expensive and I think it makes more sense to only query the `maintained` job which is our main interest for this script.

File: maintainers/scripts/haskell/hydra-report.hs
Comment (grahamc) [inline maintainers/scripts/haskell/hydra-report.hs:122]: Please attach a user-agent header with identifying information also :).

**Comment by grahamc** [comment]
I'm concerned by the volume of requests this will issue against Hydra, and the amount of load it may cause. Ideally it would execute fewer, more efficient requests to Hydra. How many people do you anticipate will run this per week?

**Comment by maralorn** [comment]
> I'm concerned by the volume of requests this will issue against Hydra, and the amount of load it may cause. Ideally it would execute fewer, more efficient requests to Hydra. How many people do you anticipate will run this per week?

Yeah. I totally understand. I am working on making the number of requests as low as possible by caching, etc.

It would probably depend a bit of the efficiency we reach in the end. But we roughly plan to run this script once, by only one person per haskell-updates merge request.

File: maintainers/scripts/haskell/hydra-report.hs
Comment (maralorn) [reply maintainers/scripts/haskell/hydra-report.hs:119]: Oh, I didn‘t know that endpoint.

**Comment by grahamc** [comment]
Can you do it in approximately two requests with:

https://github.com/NixOS/nixpkgs/pull/121713#discussion_r628754861

and

https://github.com/NixOS/nixpkgs/pull/121713#issuecomment-835375787

?

**Comment by maralorn** [comment]
> Additionally, I wonder since our main objective is getting the status of maintained packages, we should probably just check on the state of the `maintained` aggregate job instead. That would also save us from evaluating stuff locally.

@sternenseemann Well, we still kinda need the information about all broken builds for the mark-broken script.

**Comment by sternenseemann** [comment]
> Well, we still kinda need the information about all broken builds for the mark-broken script.

That's right yeah, but that script only needs to be executed *once* per merge cycle and when Hydra is finished building an evaluation, whereas we may want to update the build report multiple times.

**Comment by maralorn** [comment]
Okay. I have reduced the get-report script to only 2 queries. It runs in under 5 minutes now.

Thanks everyone for the feedback. I feel much better about this now.
While this is still not perfect I think running this script every few days by one person won’t affect hydra too much.

I have set a user agent that is hopefully easily understandable.

File: maintainers/scripts/haskell/hydra-report.hs
Comment (cdepillabout) [inline maintainers/scripts/haskell/hydra-report.hs:131]:
```suggestion
   myReq query option = responseBody <$> req GET query NoReqBody jsonResponse (header "User-Agent" "hydra-report.hs/v1 (nixpkgs;maintainers/scripts/haskell)" <> option)
```

**Comment by cdepillabout** [comment]
Thanks for fixing this up.  It of course runs significantly faster for me now.

I think this looks pretty good.  Probably ready for being merged?

This definitely isn't something that important, but if I could make one suggestion, it would be to change the package names into a link instead of (or in addition to) the build status icons:

image

It wasn't immediately obvious that I could click on the build status icons to jump to the job on Hydra.  I think if the package name was instead a link, it would be more obvious that you can click on it.

**Comment by maralorn** [comment]
Of course I can make the package name a link. The question is which one?

**Comment by cdepillabout** [comment]
> Of course I can make the package name a link. The question is which one?

Ah, I now see that each platform is a different link (which makes sense).

In that case, I agree that the current way it works is probably the best.

**Comment by maralorn** [comment]
I have added one other maintainer script to update the mark-broken list. 

Maybe someone wants to have a quick glance at that before I merge?

An example generated by it has been pushed to haskell-updates.

**Comment by maralorn** [comment]
Pushed onto haskell-updates.

**Pull request closed without merging**
